src/components/shared/TreeView.js

```diff
@@ -51,42 +51,109 @@ export type Column<DisplayData: Object> = {|
   +fixedColumns: Column<DisplayData>[],
   +mainColumn: Column<DisplayData>,
+  +viewOptions: TableViewOptions,
+  // called when the users moves the divider right of the column,
+  // passes the column index and the start x coordinate
+  +onColumnWidthChangeStart: (number, CssPixels) => void,
+  +onColumnWidthReset: (number) => void,
 |};
 
-const TreeViewHeader = <DisplayData: Object>({
-  fixedColumns,
-  mainColumn,
-}: TreeViewHeaderProps<DisplayData>) => {
-  if (fixedColumns.length === 0 && !mainColumn.titleL10nId) {
-    // If there is nothing to display in the header, do not render it.
-    return null;
-  }
-  return (
-    <div className="treeViewHeader">
-      {fixedColumns.map((col) => (
+class TreeViewHeader<DisplayData: Object> extends React.PureComponent<
+  TreeViewHeaderProps<DisplayData>
+> {
+  _onDividerMouseDown = (event: SyntheticMouseEvent<>) => {
+    this.props.onColumnWidthChangeStart(
+      // $FlowExpectError - we know that the target has this field
+      Number(event.target.attributes.columnIndex.value),
+      event.clientX
+    );
+    event.stopPropagation();
+  };
+
+  _onDividerDoubleClick = (event: SyntheticMouseEvent<>) => {
+    this.props.onColumnWidthReset(
+      // $FlowExpectError - we know that the target has this field
+      Number(event.target.attributes.columnIndex.value)
+    );
+    event.stopPropagation();
+  };
+
+  render() {
+    const { fixedColumns, mainColumn, viewOptions } = this.props;
+    const columnWidths = viewOptions.fixedColumnWidths;
+    if (fixedColumns.length === 0 && !mainColumn.titleL10nId) {
+      // If there is nothing to display in the header, do not render it.
+      return null;
+    }
+    return (
+      <div className="treeViewHeader">
+        {fixedColumns.map((col, i) => {
+          const style =
+            columnWidths || col.initialWidth
+              ? {
+                  width:
+                    (columnWidths ? columnWidths[i] : col.initialWidth) +
+                    (col.headerWidthAdjustment || 0),
+                }
+              : {};
+          return (
+            <>
+              <PermissiveLocalized
+                id={col.titleL10nId}
+                attrs={{ title: true }}
+                key={col.propName}
+              >
+                <span
+                  style={style}
+                  className={`treeViewHeaderColumn treeViewFixedColumn ${col.propName}`}
+                ></span>
+              </PermissiveLocalized>
+              {col.hideDividerAfter !== true ? (
+                <span
+                  key={col.propName + 'Divider'}
+                  className="treeViewColumnDivider"
+                  onMouseDown={
+                    col.resizable ? this._onDividerMouseDown : undefined
+                  }
+                  onDoubleClick={
+                    col.resizable ? this._onDividerDoubleClick : undefined
+                  }
+                  columnIndex={i}
```

Comment:
Another option (possibly even simpler) is to make `_onDividerMouseDown` and `_onDividerDoubleClick` take the number directly:

```js
                  onMouseDown={
                    col.resizable ? () => this._onDividerMouseDown(i) : undefined
                  }
                  onDoubleClick={
                    col.resizable ? () => this._onDividerDoubleClick(i) : undefined
                  }
```
(I removed the `event` parameter too)
Then you don't even need to add an extra attribute.

In terms of performance it's OK to use an arrow function as a parameter here, because this is a DOM Element, not a React Pure Component.

Both options (with data attribute / with arrow functions) would work for me.